### DISCUSSION TITLE
GraphQL API Pagination

### DISCUSSION THREAD

**Discussion by tartarJR** [community/community#24414] [Apps, API and Webhooks]
<p>I have a GraphQL query to search repositories and I want to paginate my result.</p>
<pre><code>search(first: $first, after:$cursor, query: $query, type: $type) {
    edges {
        node {
            ... on Repository {
                name
                description
                forkCount
                owner {
                    avatarUrl
                }
            }
        }
    }
    pageInfo {
        hasNextPage
        endCursor
    }
}
</code></pre>
<p>The query works fine when provided with a cursor. When I want to get only the first page and provide the query with a null cursor to get the first page, I get " <strong>Variable cursor of type String! was provided invalid value" </strong> error.</p>
<p>How can I retrieve the first page without having to write another query with no cursor parameter?</p>
<p>Any help would be appreciated.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/tartarjr">@tartarjr</a>,</p>
<p>Thanks for being here! If I’m not mistaken you’d have to send the initial request again because no variables are specified :frowning: Let me do a little digging on my end and see if there are any alternatives. </p>

**Comment by tartarJR** [comment]
<p>Hi Andrea </p>
<p>Thank you for your help. Please let me know if you can find any alternatives.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hey <a class="mention" href="/tartarjr">@tartarjr</a>,</p>
<p>Thanks for your patience, looks like the query works fine on our end take a look:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/bc1ab1935ea87d5c70c3dc52d934512a26ded05f.png" data-download-href="/uploads/short-url/qQ2XejoZSxCVSGYqDBy7aR55re7.png?dl=1" title="56269290-a4d12480-60f3-11e9-8725-606fbdf494ab.png"><img src="https://user-images.githubusercontent.com/20666190/181084671-9d95e74d-aa93-4a27-aaca-3a0567d4a90f.png" alt="56269290-a4d12480-60f3-11e9-8725-606fbdf494ab.png" data-base62-sha1="qQ2XejoZSxCVSGYqDBy7aR55re7" width="690" height="255" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/bc1ab1935ea87d5c70c3dc52d934512a26ded05f_2_690x255.png, https://aws1.discourse-cdn.com/github/optimized/2X/b/bc1ab1935ea87d5c70c3dc52d934512a26ded05f_2_1035x382.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/b/bc1ab1935ea87d5c70c3dc52d934512a26ded05f_2_1380x510.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/bc1ab1935ea87d5c70c3dc52d934512a26ded05f_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">56269290-a4d12480-60f3-11e9-8725-606fbdf494ab.png</span><span class="informations">3304×1222 389 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Would you please attempt you query again making sure to define the <code>$after</code> cursor parameter as <code>$after: String</code>. Please let me know if this does the trick, I hope this helps!</p>

**Comment by jfeltesse-mdsol** [comment, marked as answer]
<p>Just remove the <code>!</code> from String to make the variable accept either a String or a null value.</p>

**Comment by nithinstar093** [comment]
<p>Hey <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> </p>
<p>I was working on the search option to list all the available repositories using github api. The search option has an first field and has a limit of 100, iam trying to fetch 500 repositories, is there an alternative to do that ?</p>
<p>Code:</p>
<p>query {<br>
rateLimit{<br>
cost<br>
remaining<br>
resetAt<br>
}<br>
search( query: “-js created:&gt;2015-01-01 language:JavaScript”, type:REPOSITORY ,first: 500) {<br>
repositoryCount<br>
pageInfo {<br>
hasNextPage<br>
endCursor<br>
startCursor<br>
}<br>
edges {<br>
node {<br>
… on Repository {<br>
name<br>
url<br>
}<br>
}<br>
}<br>
}</p>
<p>}</p>
<p>“message”: “Requesting 500 records on the <code>search</code> connection exceeds the <code>first</code> limit of 100 records.”,<br>
“type”: “EXCESSIVE_PAGINATION”,</p>
<p>Thank you</p>
